### ISSUE TITLE
State of jasper

### ISSUE THREAD

**Issue by jubalh** [jasper-software/jasper#208]
[Debian](https://lists.debian.org/debian-release/2016/02/msg00734.html) and [Ubuntu](https://bugs.launchpad.net/ubuntu/+source/jasper/+bug/1639737) dropped jasper from their repositories in 2016 because of the security issues.
In 2018 there were [several](https://github.com/mdadams/jasper/issues/182) more CVEs discovered.
They have been submitted in one bug report and some of them as individual bugreports.

An effort was made in fixing these bug via a [pull request](https://github.com/mdadams/jasper/pull/157) which did not receive any feedback since Oct 6, 2017 to this day. I commented on it to ping the maintainer, no respone.
However one of the commits that were included in that PR were by hand again without mentioning the original commit/author: https://github.com/mdadams/jasper/commit/573a6e47d45c9166331c115e78be962061c00bca. The coresponding issue https://github.com/mdadams/jasper/issues/142 was also not closed.

Later another effort was made by [posting](https://github.com/mdadams/jasper/issues/182#issuecomment-451109657) fixes to several of the bugs here, located on gist.github.com.
AFAIK it were the patches used by the Debian LTS team.

They also received no feedback.

I asked the maintainer what was the reason for this, and he replied that he has limited time. Which is understandable of course. And that some pull requests are not real fixes but only mask the problem. Which is also fine but would be good to have commented on publicly on the pull request since now he himself didn't remember which PRs were the faulty ones.

I created pull request https://github.com/mdadams/jasper/pull/197 https://github.com/mdadams/jasper/pull/198 https://github.com/mdadams/jasper/pull/199 so that it is easier for the jasper maintainer to just merge them and not have to go through the pastes. https://github.com/mdadams/jasper/pull/200 was opened as a better solution for one of them.

CVE-2017-13748 was adressed in https://github.com/mdadams/jasper/pull/159 in 2017. CVE-2018-9055 was adressed in https://github.com/mdadams/jasper/pull/204 in 2019. Some more fixes in https://github.com/mdadams/jasper/pull/158.
Neither received any feedback.

Today there are 56 open issues, and 7 open pull requests.
openSUSE ships 2.0.16 with a few patches for CVEs on top.
Fedora ships 2.014 with some patches for CVEs but not fixing all the ones that have fixes available.
Debian and Ubuntu removed jasper.
Gentoo ships 2.0.16 with no patches.
Arch ships 2.0.16 with [one](https://git.archlinux.org/svntogit/packages.git/tree/trunk/jasper-1.900.1-fix-filename-buffer-overflow.patch?h=packages/jasper) patch.

With 2.0.16 (12 March 2019) the tarball was not put to the usual location, the changelog and [website](https://www.ece.uvic.ca/~frodo/jasper/) were not updated. Website still mentions JasPer 2.0.14 (2017) as latest release. See https://github.com/mdadams/jasper/issues/202 for the report.

Having limited time for an open source project is fine and understandable. But I think some parts of how things were handled could have been done better. I also think that it seems like at one point there were several people trying to help and contribute to jasper but stopped doing so because there was no reaction to their efforts.
I asked several months ago if it would be possible to at least *now* add more maintainers to the project. People who are still interested in helping jasper. Or assign some students from University of Victoria for this task. Even if not for active development/bug fixing then at least for sorting issues and taking care of pull requests but did not get a reply to this.

So currently openSUSE is also trying to reduce dependencies on jasper, and might [drop](https://bugzilla.suse.com/show_bug.cgi?id=1130404) jasper from its repos too.

My strategy was to try to help revive jasper, commenting on issues and creating pull requests.
Now I'm trying to convince other projects that rely on JasPer for JPEG-2000 support to switch to openjpeg which seems currently to be better maintained, so that when we remove jasper we don't loose JPEG-2000 support in those programs. Debian people did this for [DigiKam](https://bugs.kde.org/show_bug.cgi?id=364231) in the past too. I'm also changing openSUSE packages that can choose between the two to use [openjpeg](https://github.com/uclouvain/openjpeg) instead.
An alternative would be that upstream decides to add more maintainers and fixes bugs or that the community forks jasper and merges the current pull requests and tries to fix more.

I'm creating this issue in hope of finding more people interested in jasper so that we could find a solution together.

**Comment by SoapGentoo** [comment]
@jubalh I agree, and I have just started the process of dropping jasper from Gentoo. Consumers should move to openjpeg-2, which is maintained. Thanks for the wake up call.

**Comment by crocket** [comment]
@SoapGentoo But, you masked media-libs/jasper without fixing ebuilds that depend on jasper.

**Comment by jubalh** [comment]
> you masked media-libs/jasper without fixing ebuilds that depend on jasper.

Which should be reported on the Gentoo bugzilla or directly fixed in their tree. But not reported here :-)

**Comment by mdadams** [comment]
@jubalh  @SoapGentoo I am sorry for the situation with JasPer, but I have no one to help me with the project and I also have no funding for this project either.   Moreover, I do not do any research work related to JPEG 2000.   Although I have been told by many people that they find JasPer useful in their commercial products, no company wants to actually provide funding for the maintenance/development of the code.  So, I just do the best that I can (in my own extremely limited spare time), which is admittedly not very satisfactory.  But I don't see any alternatives, unless someone can provide support for this project.

**Comment by SoapGentoo** [comment]
@mdadams I understand that, but one approach would be to transfer it into a "JasPer trust" or round table, in order to increase the maintainer/gatekeeper bus factor. I commend @jubalh for trying to find a community-oriented solution at this point.

**Comment by Justinzobel** [comment]
@SoapGentoo that would be a great plan.

**Comment by SoapGentoo** [comment]
@Justinzobel Gentoo has removed jasper two weeks ago.

**Comment by Justinzobel** [comment]
@SoapGentoo an unfortunate necessity I imagine. Just a pity. I just wish someone could pick it up, fork it and provide the necessary patches as many things still rely on jasper.

**Comment by jubalh** [comment]
It's not just about picking it up. Finding solutions for some of the issues is not easy.

**Comment by Justinzobel** [comment]
Then it's about getting downstream projects moving on from abandoned software.

**Comment by jubalh** [comment]
Which is what we are doing here since a few weeks ;-)
I wrote to all downstream projects that use jasper and have packages in openSUSE.

You are late to the party ;)

**Comment by Justinzobel** [comment]
Thank you then for writing to them all.

**Comment by Justinzobel** [comment]
Do you mind linking the issues?

**Comment by jubalh** [comment]
> Do you mind linking the issues?

**Comment by jubalh** [comment]
We are making progress.

Two months ago openSUSE [removed](https://build.opensuse.org/request/show/769072) Jasper from its repositories too.

Today OpenCV merged a [patch](https://github.com/opencv/opencv/pull/16494) to use OpenJPEG instead of Jasper for JPEG2000 support.

Nevertheless I opened [another](https://github.com/mdadams/jasper/pull/216) pull request which fixes CVE-2018-9154.
There are still 9 pull requests open.

@mdadams would you consider taking a look at merging them?

**Comment by jubalh** [comment]
So I would like to try it on more time.

Do you think we can work together here to improve the state of the project?
You could [add collaborators](https://help.github.jp/enterprise/2.11/user/articles/inviting-collaborators-to-a-personal-repository/) to your repo so that more people can merge pull requests and review things. I would volunteer to do so.

If even more people would like to joind or you don't want any responsibility at all or don't want to host it on your private account we could also create an organization on GitHub and [transfer](https://help.github.jp/enterprise/2.11/user/articles/transferring-a-repository-owned-by-your-personal-account/) the repository there.

If you send us the source for the website currently hosted at https://www.ece.uvic.ca/~frodo/jasper/ we could also use [GitHub Pages](https://pages.github.com/) to host the site on GitHub.

With more people spending time on this there is a better chance that this thing keeps on going.
Like I said, I would volunteer to help, and maybe others join too.
What do you think @mdadams ?

**Comment by jubalh** [comment]
@thoger would you be interested in maintaing a fork of jasper with me? I saw that you too seem to look at old bugs. I tried to send you an email to your RH adress yesterday. Not sure whether it reached you.

**Comment by jubalh** [comment]
The organization got created. We have a repo at https://github.com/jasper-maint/jasper now.

**Comment by MaxKellermann** [comment]
Resuming the discussion from https://github.com/mdadams/jasper/pull/159#issuecomment-648166484 (because this is the right place):

@mdadams:

Our fork at https://github.com/jasper-maint/jasper/ now has fixes for **all** known JasPer vulnerabilities. And many more - I have been fuzzing JasPer for a few days, and libFuzzer has found numerous additional vulnerabilities, *all* of which I fixed. Our fork has 101 new commits since your release 2.0.16.

If you are serious about keeping the official JasPer project alive, merge our fixes into your official JasPer repository.

**Comment by jubalh** [comment]
You can merge them easily. Then push to your new organization. If you need help how to do this let us know.
In the new organization we can help sorting/commenting on in issues. Don't forget to give us push access to the repo if you intend to do so, I think that's a separate setting.

Next steps:
* Create ChangeLog (that also mentions all CVEs that have been fixed)
* Create new release

This will either be done together or in our fork, depending on the feedback/action.

**Comment by jubalh** [comment]
@mdadams another week has passed.

While I totally get that you probably are quite busy, I absolutely don't get why you won't allow the project to live without need of you immediate action.

Yes, you created the `jasper-software` organization on GitHub (after we created `jasper-maint`..) but you are the only owner of the organization. @MaxKellermann and I can't do much there. So we are totally blocked to do anything there.
Also the old repo is under your personal account, so we couldn't even transfer it. So we are blocked again..

All CVEs (and more) got fixed. Code got cleaned up pretty nicely. We tried to organize and comment on existing issues, making all the work transparent and easy to follow.

So all the technical work is done. But more organizational work is needed.
Release notes to be written, a release put out. So that finally after many years, people actually can use a safe version of JasPer again.

We would still like to help here, but you mentioned your opposition of our fork. Saying it needs to be done in a way that you feel comfortable with.

Well, I'm not sure what you are comfortable with and with what not. I only see that you cling to this project in a way blocking all progress.

Let us just take the easiest route now:
Join our `jasper-maint` organization, so you stay in control of JasPer. (Which we already sent before but the invitation expired because you didn't accept it in time)
Write a note in the Readme of your repo pointing to the new repo saying it is the new upstream.

This means the least amount of work for you. And the quickest way to a new JasPer for the rest of the world ;)

**Comment by SoapGentoo** [comment]
@mdadams please make jasper usable for the community by transferring it. Otherwise more and more distributions and packagers will remove it from their repositories.

**Comment by jubalh** [comment]
If it won't be transferred soonish we will just start creating releases (and continue maintaining) at our fork: https://github.com/jasper-maint/jasper

In that scenario openSUSE will definitely switch to that fork.

**Comment by jubalh** [comment]
@mdadams I hope you can find some time to respond.

If you won't respond within a week we will:
* Write a ChangeLog
* Create a new release (2.0.17 or 3.0.0 @MaxKellermann any preference?)
* Still use the name "jasper"
* Notify potential users about our release and project

**Comment by MaxKellermann** [comment]
I say 2.0.17, because the release is supposed to be an ABI-compatible drop-in for 2.0.16.
JasPer's ABI is very less-than-optimal because it uses the notoriously slow (and badly named) `int_fast32_t`. Breaking the ABI would have lots of advantages (like https://github.com/jasper-maint/jasper/commit/52cb87e908b8a71f5e8fd0b5ed5260829e4335a6). Also, JasPer exposes too many internals in its API/ABI. I have lots of ideas and good reasons for breaking the ABI, but I'm not sure if we'll ever do that. There may never be JasPer version 3 with an optimized ABI.

**Comment by jubalh** [comment]
> I say 2.0.17, because the release is supposed to be an ABI-compatible drop-in for 2.0.16.

Agreed. Though in the past it seems .so name bump and ABI incompatibly wasn't related (although of course it should) ;) https://abi-laboratory.pro/index.php?view=timeline&l=jasper

**Comment by mdadams** [comment]
> I say 2.0.17, because the release is supposed to be an ABI-compatible drop-in for 2.0.16.
> JasPer's ABI is very less-than-optimal because it uses the notoriously slow (and badly named) `int_fast32_t`. Breaking the ABI would have lots of advantages (like [jasper-maint/jasper@52cb87e](https://github.com/jasper-maint/jasper/commit/52cb87e908b8a71f5e8fd0b5ed5260829e4335a6)). Also, JasPer exposes too many internals in its API/ABI. I have lots of ideas and good reasons for breaking the ABI, but I'm not sure if we'll ever do that. There may never be JasPer version 3 with an optimized ABI.

@MaxKellermann and @jubalh 
I have manually merged all of the changes from your repository (as of earlier today) into a new jasper-maint branch, with two minor modifications (namely, the README and reviving one deleted file).  I also tagged the commit with version-2.0.17 as Max suggested.  If there is a more optimal way that I could have used to pickup your changes with some additional metadata but still make a few modifications before propagating the result to my repository, please let me know and I can do that.  I wasn't sure if I could do this by forking and merging and still apply some changes before propagating the result to the original repository.  So, I did what I knew would work so as not to delay things more.   I also sent an email to each of you with some additional details and discussion that is probably better handed via email.

**Comment by jubalh** [comment]
@mdadams I answered to all your mails now but didn't get any reply.
So I'll move the discussion here again just to have it transparent for everybody.

If we want to merge the project together we should be quick. Otherwise people create PRs to one or the other and things might diverge..

For example you received: https://github.com/jasper-software/jasper/pull/218
While we received: https://github.com/jasper-maint/jasper/pull/45

And since you last reviewed our changes I also merged: https://github.com/jasper-maint/jasper/pull/44

So please let's hurry to come to a conclusion together.

Like I wrote in the mail your changes in https://github.com/jasper-software/jasper/commit/3b9c3076096652e114086a831a214bddf7f60b78 are not good.
They add all our changes as *one single commit*. But we had something over 100 commits, which contained also detailed descriptions in the commit messages and references to issues.

Also then just tagging a release without clearing up the state of the project and not having a proper changelog is also not good in my opinion.

You mentioned that you merged all our changes but two into that commit. It's very hard to find out *which* these were.
For this reason you should have merged so that each commit stays one commit, and then did two revert commits for the commits which you don't agree with.

So please now do this.

* merge our changes
* git revert cbcaf65cc667fb23d65a22526b7a42a4761303eb (and add a commit message describing why its reverted)
* git revert (second thing you didnt like)
* push this to master

Then we can discuss how we will go about the ChangeLog. Since this release fixes *all* known CVEs for JasPer it's important to highlight this to the user.

If you are not that familiar with git, please tell me the two commits you don't like. Give me write access to this repo and I'll do it for you.
But let's do it *soon*.

**Comment by theta682** [comment]
@jubalh, @MaxKellermann: I see the version 2.0.17 was released (from a local jasper-maint branch, not master) with all commits from jasper-maint/jasper repository and two commits on top. Now we have official release which includes all your hard work. 
However, I agree with you that this project needs a collaborative team that can fast fix security vulnerabilities. I like @MaxKellermann ideas to improve performance.

**Comment by theta682** [comment]
@jubalh, @MaxKellermann thank you for great work!

**Comment by osamu620** [comment]
In terms of compliance with the international standard, PR https://github.com/jasper-maint/jasper/pull/45 is very important. Now the JPEG group is looking for a long-term way how to maintain reference software packages. Depending on future discussions and if Jasper does not comply with the standard,  it may become necessary to drop Jasper from the official reference software.

**Comment by jubalh** [comment]
@mdadams good! The work on the now new `jasper-maint` branch (apparently the one where you tagged last is now `jasper-maint-old`) looks much better! Up to https://github.com/jasper-software/jasper/commit/50bd3ad7292d24ba14e8b3f9df35bf81eca2b1b1.

That last commit could have been done in two commits and using git revert. But its no big deal.

I think we should wait with tagging until we are sure all is fine in general :-)

>  I see the version 2.0.17 was released (from a local jasper-maint branch, not master) with all commits from jasper-maint/jasper repository and two commits on top.

I also think we should tag on master only :-)

> In terms of compliance with the international standard, PR jasper-maint/jasper#45 is very important.
Yep! We won't forget to review/merge it.

> Now the JPEG group is looking for a long-term way how to maintain reference software packages. Depending on future discussions and if Jasper does not comply with the standard, it may become necessary to drop Jasper from the official reference software.

I think slowly we are gaining ground and JasPer will survive. Which means we will try to comply with the standard and will review pull requests (once all the current confusion has settled). So no worries :-)

**Comment by jubalh** [comment]
I merged all our commits (as of today) from https://github.com/jasper-maint/jasper to the master branch of https://github.com/jasper-software/jasper.

I then reverted the note about the fork in the README: https://github.com/jasper-software/jasper/commit/f74f8827e0b4b0ce0d2cd1d5396ea7e9425be6bd
Updated the links in the README from mdadams to jasper-software https://github.com/jasper-software/jasper/commit/0663cdc9ce0f9ca479391a49e34187307ff6c7d9

Added a (not the nicest) changelog: https://github.com/jasper-software/jasper/commit/eaa7abb7faa8336380e3b53012300d2dc2426f14 containing mentions of all the CVEs we fixed.
Certainly this can be improved but I hope it is okay for now.

And I bumpbed the version to 2.0.19 and tagged it: https://github.com/jasper-software/jasper/commit/7d8cfd8ac16d1af9b51e5ccd781e898f0fbf57cc

My goal is now to move https://github.com/jasper-maint/jasper/pull/45 from jasper-maint to jasper-software.
Then hoping @MaxKellermann and @mdadams can review this PR. So it can be part of the next release.

Then I plan to archive/disable pull requests/issues on jasper-maint and add a note that we merged here in jasper-software with the original author. I hope this reduced confusion for users.

https://repology.org/project/jasper/versions shows that distributions also are confused already.
In the Arch AUR they use the 2.0.17 already even though it was tagged two times on different branches/commits. Not good.

Slackware and CRUX already use a 2.0.18.

We need to align all this and make it more visible to users whats going on behind the scenes.

So I took the liberty to do the changes mentioned above.
From here we can continue to work on JasPer together in one repository under the lead of @mdadams.

@mdadams you told us you wanted to review our changes first before merging them on master. But the confusion that we now have needed correction. This is why I did the above steps. Otherwise we end up in a huge confused mess.

You can still review all our changes and then use `git rever commit-hash` to rever what you don't like. Preferable only after we discussed it together. You can then still create a new release (tagged on master) that is the way you want it.
But having a 2.0.17 with our changes out in public for people to use it and not having them on master because you first want to review it, makes little sense to me.

I hope now all projects switch to 2.0.19 and we can continue to work all together on JasPer.

**Comment by jubalh** [comment]
@thoger maybe you can inform the Fedora/RedHat maintainer
@apoleon AFAIK you worked on JasPer for Debian LTS, maybe you want to add it back into Debian?
@ncopa maybe you want to update the Alpine package
@anthraxx maybe you want to update the Arch package
@SoapGentoo I created https://bugs.gentoo.org/734284

[Submission](https://build.opensuse.org/request/show/823122) to add JasPer 2.0.19 again to the official openSUSE repo.

I [informed](https://bugs.kde.org/show_bug.cgi?id=364231#c8) Digikam that there is no need to replace JasPer anymore.

**Comment by SoapGentoo** [comment]
@jubalh thanks, I'll look into readding jasper back to Gentoo.

**Comment by ncopa** [comment]
@jubalh Impressive work you have done here. Thank you!

**Comment by mdadams** [comment]
@jubalh and @MaxKellermann:
Thank you for your hard work, Michael and Max.

I'm sorry about the tag issue.  This was my fault.  Since 2.0.17 was used to tag different commits at different points in time, it might be best to delete this tag because any references to it might not really be referencing what people think it references.  At least if the tag were deleted, people would know that something is wrong, rather than unknowingly referencing the wrong commit.

**Comment by mdadams** [comment]
@jubalh:  Okay, I understand your rationale for merging the changes now.  It makes sense, given the circumstances.  I'm sorry for the botched 2.0.17 tag.  I can review the merged changes after the fact.  Perhaps, merging sooner will also help to consolidate the various JasPer efforts more quickly, which would only be a good thing.

**Comment by Justinzobel** [comment]
Oh damn I missed this issue's birthday, happy birthday issue for 26 days ago!

**Comment by jubalh** [comment]
All our fixes and improvements have been merged from jasper-maint/jasper into jasper-software/jasper.

The last pull request moved from https://github.com/jasper-maint/jasper/pull/45 to https://github.com/jasper-software/jasper/pull/221  for further review and discussion.

I put a note in the about section at https://github.com/jasper-maint/jasper and archived the repository.

Current state of this project:
* all known CVEs fixed.
* 3 people with commit rights (bottleneck solved)
* more contributors
* distributions notified

**Comment by anthraxx** [comment]
The 2.0.17 is already release in some distros. Please don't simply delete a tag or do things like re-tagging. If you want to correct this please release an appropriate followup version 2.0.18.

**Comment by jubalh** [comment]
@anthraxx maybe you can start by at least reading the last couple of comments in this thread...

**Comment by anthraxx** [comment]
@jubalh I'm following the whole thread in fact since it has been born. I'm just saying its in my humble opinion not a good idea to delete tags. It can be annotated in the release description that its a broken version or something.
The last sentences clearly say what i tried to summarize, please never re-tag tags as you yourself said "it was tagged two times on different branches/commits".

**Comment by jubalh** [comment]
@anthraxx You are saying what we have said since many comments. Plus the addition to not delete the tag.
And then you say we should create a follow up tag "2.0.18" when we have already released 2.0.19. The reason for 19 was also discussed in this thread.

**Comment by anthraxx** [comment]
@jubalh I think you are not  really reading it in a way it was meant -- and/or maybe I just wrote it badly in a non unambiguous way. That was meant as an example what should be done instead of re-tagging. Yes, it may have been just a +1, but I wanted to add a written feedback from another distro packager, consider it more like a +1 ACK but with the difference I would not delete tags even if they are faulty.
I'm following here because we were also in the middle of nuking jasper out of existence in our distro

**Comment by jubalh** [comment]
Buildroot has been updated to 2.0.19 too: https://github.com/buildroot/buildroot/commit/d0f7b241d4619603adca32fd03d2687126e9e7e3
